src/components/Table/TableBase.tsx

```diff
@@ -404,16 +407,50 @@ export class TableBase<T> extends React.Component<
 		const identifier = item[rowKey];
 		const newIsChecked = !this.getCheckedRowIdentifiers().has(identifier);
 
-		const oldCheckedItems = this.props.checkedItems || this.state.checkedItems;
-		const checkedItems = newIsChecked
-			? oldCheckedItems.concat(item)
-			: (oldCheckedItems ?? []).filter((x) => x[rowKey] !== identifier);
+		let checkedItems = this.props.checkedItems || this.state.checkedItems;
+
+		let keysToFind = [identifier];
+		if (
+			this.state.lastSelected !== null &&
+			(e.nativeEvent as MouseEvent).shiftKey
+		) {
+			keysToFind.push(this.state.lastSelected);
+		}
+		let index = 0;
+		while (keysToFind.length) {
+			let foundKey = false;
+			if (keysToFind.includes(data[index][rowKey])) {
+				keysToFind = keysToFind.filter(
+					(keyToFind) => keyToFind !== data[index][rowKey],
+				);
+				foundKey = true;
+			}
+			if (
+				foundKey ||
+				((e.nativeEvent as MouseEvent).shiftKey &&
+					keysToFind.length === 1 &&
+					this.state.lastSelected !== null)
+			) {
+				const itemIsChecked =
+					checkedItems.filter((item) => item[rowKey] === data[index][rowKey])
+						.length === 0;
+				if (newIsChecked && itemIsChecked) {
+					checkedItems = checkedItems.concat(data[index]);
+				} else if (!newIsChecked && !itemIsChecked) {
+					checkedItems = checkedItems.filter(
+						(item) => item[rowKey] !== data[index][rowKey],
+					);
```

Comment:
Let's avoid the `O(N^2)`:
```suggestion
					checkedItemsMap.delete(data[index][rowKey]);
```